openmdao/core/driver.py

```diff
@@ -1061,20 +1073,39 @@ def record_iteration(requester, prob, case_name):
     filt = requester._filtered_vars_to_record
     model = prob.model
 
+    inputs, outputs, residuals = model.get_nonlinear_vectors()
+    vec_name = 'nonlinear'
+
     parallel = requester._rec_mgr._check_parallel() if model.comm.size > 1 else False
 
     outs = model._retrieve_data_of_kind(filt, 'output', 'nonlinear', parallel)
     ins = model._retrieve_data_of_kind(filt, 'input', 'nonlinear', parallel)
 
-    data = {
-        'output': outs,
-        'input': ins
-    }
+    data = {'input': ins, 'output': outs, 'residual': {}}
+
+    if requester.recording_options['record_residuals'] and residuals._names:
+        data['residual'] = model._retrieve_data_of_kind(
+            filt, 'residual', vec_name, parallel)
+
     from openmdao.core.problem import Problem
-    if isinstance(requester, Problem) and requester.recording_options['record_derivatives'] and \
-            prob.driver._designvars and prob.driver._responses:
-        totals = requester.compute_totals(return_format='flat_dict_structured_key')
-        data['totals'] = totals
+    if isinstance(requester, Problem):
+        inputs, outputs, residuals = model.get_nonlinear_vectors()
+        vec_name = 'nonlinear'
+
+        discrete_inputs = model._discrete_inputs
+        discrete_outputs = model._discrete_outputs
+        filt = requester._filtered_vars_to_record
```

Comment:
The above lines appear to be unnecessary? 
- 1092-1093 repeat 1076-1077
- 1095-1096 are unused?
- 1097 repeats 1073

Also, if requester is Problem lines 1081-1088 are executed unnecessarily?

Reply:
let me look! I wouldn't doubt this needs work